### PR TITLE
action recognition 

### PR DESCRIPTION
**Pull request by jiata**
### Description
- [x] update video dataset and learner with new visualization and evaluation functionality
- [x] update 01_training_intro notebook with new text/explanations 
- [x] created and uploaded new dataset of 60 vids to our cvbp storage. each vid is from 5-7 seconds each, downsized to 400X400px, 2 classes (pouring/opening)
- [ ] tests for all of the above and past action recognition code

> note: tests aren't done (and i haven't included all of them in this pr), but feel free to review other aspects

### Checklist:

- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [x] This PR is being made to `staging` and not `master`
- [x] I will squash merge this PR into `staging`

### CODE REVIEW

File: scenarios/action_recognition/02_training_hmdb.ipynb [hunk 1]
```diff
@@ -34,7 +34,7 @@
   },
```
Comment (PatrickBue) [inline scenarios/action_recognition/02_training_hmdb.ipynb:34]: <p>Good notebook.</p><p>Could point out here (to make it even more explicit) that the paper reports an accuracy of XX% on the popular research benchmark dataset HMDB-51, and with this notebook we achieve XX%, showing that our implementation is indeed correct.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/535/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Comment by PatrickBue** [comment]
@jiata : just had a look at the 01 and 02 notebooks and left comments using ReviewNB. Looks great, only minor comments, the main one to show how to score a single video. Next will be looking at the other notebooks, and then the code.

**Comment by PatrickBue** [comment]
@jiata I cannot leave comments for the 00 and 10 notebook using ReviewNB, hence here now.

10 notebook (mainly minor points):
- Consider adding a couple of sentences at the very start of the notebook what the notebook is for. Ie. the notebook is showing what different transformations look like, to help decide what transformations to use during training.
- The various "<Figure size 432x288 with 0 Axes>" shouldn't be there if possible, also is there an easy way to suppress the warnings?
- [minor] might look nicer if changing font to be larger for some of the subsections, for e.g. "Horizontal flipping" to "###Horizontal flipping".

**Comment by PatrickBue** [comment]
The 00 notebook feels a bit unfinished.
- I'd love to move all the comments how to set up / use a webcam into a FAQ under image classification and then remove the explanation from all our 00 notebooks. But that's out of scope for this PR of course. But maybe we can start just with this notebook?
- Add reference to classification/FAQ.md which explains what to do when the widgets are not showing.
- Don't think MEAN and STD should be defined inside the notebook. They already are defined in dataset.py
- Not sure why we need to define both IM_SCALE and INPUT_SIZE in the 00 notebook. Guess that level of detail could be hidden? We only ever expose IM_SIZE or similar to the user for detection, classification, etc.
- This is using the "old" way to load the model, but should rather mimic how the 01 notebook works, ie using the VideoLearner().
- Functions like predict() should likely not be defined in the notebook, but rather similar to object detection a function in the learner: def predict(). 
- Generally, there seems to be a lot of code in this notebook. I wonder how much of that is a general utility and should be in utils.

**Comment by jiata** [comment]
@PatrickBue updated 01, 02 and 10. I'll fix the following in a separate pr.
- your comments about notebook 00
- adding scoring function to 01 notebook

File: scenarios/action_recognition/02_training_hmdb.ipynb (hunk 1 above)
Comment (PatrickBue) [inline scenarios/action_recognition/02_training_hmdb.ipynb:34]: <p>[minor] could put download into utils_cv.common</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/535/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (PatrickBue) [inline scenarios/action_recognition/02_training_hmdb.ipynb:34]: <p>This needs a little more explanation what is going on and what these itemized items mean. Plus a comparison between the reported accuracy, and the accuracy we achieve would be great. Looks like we only get 58.5% accuracy versus the reported 74.5% ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/535/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

File: scenarios/action_recognition/00_webcam.ipynb [hunk 2]
```diff
@@ -1,5 +1,14 @@
 {
```
Comment (PatrickBue) [inline scenarios/action_recognition/00_webcam.ipynb:1]: <p>[minor] not sure if the sentence is ending here "By default, our VideoLearner will load.."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/535/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (PatrickBue) [inline scenarios/action_recognition/00_webcam.ipynb:1]: <p>[very minor] maybe only show eg 10 to save some space</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/535/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (PatrickBue) [inline scenarios/action_recognition/00_webcam.ipynb:1]: <ul><li>This and below looks great!</li><li>Does the generated video look roughly like the Jun Ki's in terms of accuracy?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/535/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

File: tests/unit/action_recognition/test_action_recognition_notebooks.py [hunk 3]
```diff
@@ -39,8 +44,8 @@ def test_01_notebook_run(action_recognition_notebooks):
         OUTPUT_NOTEBOOK,
         parameters=dict(
             PM_VERSION=pm.__version__,
-            MODEL_INPUT_SIZE=4,
-            EPOCHS=4,
+            MODEL_INPUT_SIZE=8,
+            EPOCHS=2,
```
Comment (PatrickBue) [inline tests/unit/action_recognition/test_action_recognition_notebooks.py:48]: Not sure how slow action recognition training/inference exactly is, hence: are these notebook tests fast enough?

File: utils_cv/action_recognition/model.py [hunk 4]
```diff
@@ -3,10 +3,9 @@
 
 from collections import OrderedDict
 import os
-import time
 import warnings
 import numpy as np
-from typing import Union, Dict, Tuple, Any
+from typing import Union, Dict, Tuple, Any, List, Callable
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:8]: [minor] imports not fully alphabetically sorted

File: utils_cv/action_recognition/model.py [hunk 5]
```diff
@@ -61,16 +72,21 @@ def __init__(
...
         # set num classes
         self.num_classes = num_classes
 
-        # set empty - populated when fit is called
-        self.results = []
+        if dataset:
+            self.dataset = dataset
+            self.sample_length = self.dataset.sample_length
+        else:
+            self.sample_length = 8 if sample_length <= 8 else 32
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:85]: Maybe add an assert here to make sure the user specifies either 8 or 32, rather than overwriting user input silently

File: utils_cv/action_recognition/model.py [hunk 6]
```diff
@@ -123,9 +134,11 @@ def freeze(self) -> None:
             param.requires_grad = True
 
     def unfreeze(self) -> None:
+        """ todo """
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:137]: Left-over text "todo". Make sure please all your functions are commented with a description of each inputs.

File: utils_cv/action_recognition/model.py [hunk 7]
```diff
@@ -475,19 +488,190 @@ def evaluate(
...
         )
         print(
             "Clip prediction accuracy =",
-            accuracy_score(ret["clip_trues"], ret["clip_preds"]),
+            round(accuracy_score(ret["clip_trues"], ret["clip_preds"]), 2),
         )
         return ret
 
+    def _predict(self, frames, transform):
+        """ Todo """
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:505]: todo

File: utils_cv/action_recognition/model.py [hunk 8]
```diff
@@ -475,19 +488,190 @@ def evaluate(
...
+
+    def _filter_labels(
+        self,
+        id_score_dict: dict,
+        labels: List[str],
+        threshold: float = 0.0,
+        target_labels: List[str] = None,
+        filter_labels: List[str] = None,
+    ) -> Dict[str, int]:
+        """ Todo """
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:522]: todo

File: utils_cv/action_recognition/model.py [hunk 9]
```diff
@@ -475,19 +488,190 @@ def evaluate(
...
+            # Plot final results nicely
+            update_println(println)
+            scores_sum -= scores_cache.popleft()
+
+        # Inference done. Ready to run on the next frames.
+        window.popleft()
+        if is_ready:
+            is_ready[0] = True
+
+    def predict_video(
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:594]: Should we split this function into two:
- One that provides the current functionality including the IPython display in a notebook
- One that takes a video and returns the prediction scores, without any plotting/printing/etc (ie how it would be used during a call to a deployed API)

File: utils_cv/action_recognition/model.py [hunk 10]
```diff
@@ -317,19 +371,307 @@ def train_an_epoch(
...
+                                labels,
+                                target_labels,
+                                transforms,
+                                update_println,
+                            ),
+                        ).start()
+
+                # Show video preview
+                f = io.BytesIO()
+                im = Image.fromarray(frame)
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:667]: [minor] Fyi, the widget is flickering heavily on my machine. Simply resizing the frame here before calling d_video.update() seems to help significantly with that.
               w,h = frame.shape[0], frame.shape[1]
                scale = 300.0/max(w,h)
                w = round(w*scale)
                h = round(h*scale)
                im = im.resize((h,w))

File: utils_cv/action_recognition/model.py [hunk 11]
```diff
@@ -276,14 +328,16 @@ def train_an_epoch(
...
             dl = data_loaders[phase]
 
+            # collect metrics
             batch_time = AverageMeter()
             losses = AverageMeter()
             top1 = AverageMeter()
             top5 = AverageMeter()
 
-            end = time.time()
+            end = time()
```
Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:340]: [minor] Could put code below under the "for step, (inputs,..." loop. Otherwise the 02 notebook will not show any sign of life in a long time, plus this does not show in the 01 notebook.
                if step % 10 == 0:
                    print(f"   Phase {phase}: batch {step} of {len(dl)}")

Comment (PatrickBue) [inline utils_cv/action_recognition/model.py:340]: [minor] Could put code below under the "for step, (inputs,..." line. Otherwise the 02 notebook will not show any sign of life in a long time, plus this does not show in the 01 notebook.
                if step % 10 == 0:
                    print(f"   Phase {phase}: batch {step} of {len(dl)}")

Comment (jiata) [reply utils_cv/action_recognition/model.py:340]: done

File: utils_cv/action_recognition/model.py (hunk 10 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:667]: done

File: utils_cv/action_recognition/model.py (hunk 9 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:594]: `predict_frames` is the function that basically returns the scores.
`predict_video` handles the current functionality of processing the video and displaying the ipython disply in the notebook. 

I moved some of the model-scoring specific code into `predict_frames` to make this distinction clearer

Currently, `predict_video` called `predict_frames`

File: utils_cv/action_recognition/model.py (hunk 8 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:522]: done

File: utils_cv/action_recognition/model.py (hunk 7 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:505]: done

File: utils_cv/action_recognition/model.py (hunk 6 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:137]: done

File: utils_cv/action_recognition/model.py (hunk 5 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:85]: done

File: utils_cv/action_recognition/model.py (hunk 4 above)
Comment (jiata) [reply utils_cv/action_recognition/model.py:8]: done

File: tests/unit/action_recognition/test_action_recognition_notebooks.py (hunk 3 above)
Comment (jiata) [reply tests/unit/action_recognition/test_action_recognition_notebooks.py:48]: yeah they aren't too slow at 2 and 8 respectively

File: scenarios/action_recognition/00_webcam.ipynb (hunk 2 above)
Comment (jiata) [reply scenarios/action_recognition/00_webcam.ipynb:1]: should be

Comment (jiata) [reply scenarios/action_recognition/00_webcam.ipynb:1]: done

Comment (jiata) [reply scenarios/action_recognition/00_webcam.ipynb:1]: done

File: scenarios/action_recognition/02_training_hmdb.ipynb (hunk 1 above)
Comment (jiata) [reply scenarios/action_recognition/02_training_hmdb.ipynb]: done

Comment (jiata) [reply scenarios/action_recognition/02_training_hmdb.ipynb]: done
